examples/softirq-latency.bpf.c

```diff
@@ -0,0 +1,117 @@
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+// Loosely based on https://github.com/xdp-project/xdp-project/blob/master/areas/latency/softirq_net_latency.bt
+
+// 20 buckets for latency, max range is 0.5s .. 1.0s
+#define MAX_LATENCY_SLOT 21
+
+struct softirq_latency_key_t {
+    u32 kind;
+    u64 slot;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirqs_raise SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirqs_entry SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, struct softirq_latency_key_t);
+    __type(value, u64);
+} softirq_entry_latency_seconds SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, struct softirq_latency_key_t);
+    __type(value, u64);
+} softirq_service_latency_seconds SEC(".maps");
+
+SEC("tp_btf/softirq_raise")
+int BPF_PROG(softirq_raise, unsigned int vec_nr)
+{
+    u64 ts = bpf_ktime_get_ns();
+    bpf_map_update_elem(&softirqs_raise, &vec_nr, &ts, BPF_ANY);
+    return 0;
+}
+
+SEC("tp_btf/softirq_entry")
+int BPF_PROG(softirq_entry, unsigned int vec_nr)
+{
+    u64 delta_us, latency_slot, *tsp, ts = bpf_ktime_get_ns();
+    struct softirq_latency_key_t latency_key = {};
+
+    tsp = bpf_map_lookup_elem(&softirqs_raise, &vec_nr);
+    if (!tsp) {
+        return 0;
+    }
+
+    delta_us = (ts - *tsp) / 1000;
```

Comment:
It's possible for `*tsp` from `softirq_raise` to be newer than `ts` from `softirq_entry`:

```c
    if (*tsp > ts) {
        bpf_printk("re-raise after %dns", *tsp - ts);
        return 0;
    }
```

```
          <idle>-0       [010] d.s4. 336314.801633: bpf_trace_printk: re-raise after 500ns
          <idle>-0       [009] d.s4. 336314.806655: bpf_trace_printk: re-raise after 740ns
           <...>-82187   [038] d.s4. 336314.808602: bpf_trace_printk: re-raise after 1390ns
          <idle>-0       [012] d.s4. 336314.829625: bpf_trace_printk: re-raise after 1170ns
          <idle>-0       [012] d.s4. 336314.904658: bpf_trace_printk: re-raise after 820ns
          <idle>-0       [035] d.s4. 336314.919042: bpf_trace_printk: re-raise after 1510ns
          <idle>-0       [002] d.s4. 336314.936068: bpf_trace_printk: re-raise after 1290ns
          <idle>-0       [001] d.s4. 336314.937639: bpf_trace_printk: re-raise after 1360ns
          <idle>-0       [001] d.s4. 336314.939628: bpf_trace_printk: re-raise after 650ns
          <idle>-0       [008] d.s4. 336314.988633: bpf_trace_printk: re-raise after 470ns
```

We need to guard against that, otherwise metrics produce incorrect values.

I think the softirq can be re-raised after this call:

* https://github.com/torvalds/linux/blob/v6.6-rc6/kernel/softirq.c#L535

Which happens before the previously raised softirq is processed.

Reply:
This can be observed with perf:

```
$ sudo perf record --cpu 6 -e irq:softirq_raise,irq:softirq_entry,irq:softirq_exit -- sleep 1
$ sudo perf script | fgrep NET_RX
```

A typical non-racy `raise`-`entry`-`exit` cycle:

```
         swapper     0 [006] 343483.505449: irq:softirq_raise: vec=3 [action=NET_RX]
         swapper     0 [006] 343483.505451: irq:softirq_entry: vec=3 [action=NET_RX]
         swapper     0 [006] 343483.505458:  irq:softirq_exit: vec=3 [action=NET_RX]
```

A problematic `raise`-`entry`-`raise`-`exit`:

```
         swapper     0 [006] 343483.519984: irq:softirq_raise: vec=3 [action=NET_RX]
         swapper     0 [006] 343483.519986: irq:softirq_entry: vec=3 [action=NET_RX]
         swapper     0 [006] 343483.520025: irq:softirq_raise: vec=3 [action=NET_RX]
         swapper     0 [006] 343483.520026:  irq:softirq_exit: vec=3 [action=NET_RX]
```

One can also find an even more problematic `raise`-`raise`-`entry`-`exit`:

```
$ sudo perf script -f | fgrep NET_RX | awk '{ if (raised == 2) { print $0 } if ($5 == "irq:softirq_raise:") { raised += 1 } else { raised = 0 } }'
         swapper     0 [006] 343482.651614: irq:softirq_entry: vec=3 [action=NET_RX]
```

```
$ sudo perf script -f | fgrep NET_RX | fgrep 343482.651614 -B2 -A1
         swapper     0 [006] 343482.651551: irq:softirq_raise: vec=3 [action=NET_RX]
         swapper     0 [006] 343482.651590: irq:softirq_raise: vec=3 [action=NET_RX]
         swapper     0 [006] 343482.651614: irq:softirq_entry: vec=3 [action=NET_RX]
         swapper     0 [006] 343482.651630:  irq:softirq_exit: vec=3 [action=NET_RX]
```